smolyay/surrogate.py

```diff
@@ -117,430 +72,694 @@ def data(self):
+    :attr:`index_combinations` describes the combination of basis function
+    used to construct the terms of the surrogate, where each row is a term
+    represented by a list of size ``num_dimensions`` that give the index of
+    the basis functions from the :class:BasisFunctionSets in `basis_sets`
+    that make up a given term.
+    :attr:`coefficients` is the list of coefficients for each term in the
+    surrogate equation. These cefficients are determined in :meth:`fit`.
+    :meth:`train`, generates a trained surrogate model given a function and
+    the points to sample at.
+    :meth:`fit` computes the coefficients based on a set of data
+    (function values at transformed grid points).
+    Once the surrogate is constructed, one can evaluate the surrogate
+    through :meth:`predict(x)` and its gradient can be evaluated
+    through :meth:`gradient(x)`.
+    :meth:`_create_terms`, is an abstract method for generating the terms
+    described in :attr:`index_combinations`.
+
+    Parameters
+    ----------
+    domain: numpy.ndarray
+        the domain of the surrgoate
+
+    basis_sets: list of :class:BasisFunctionSet
+        the set of basis functions used to combine terms
+
+    alpha: float, default 1e-10
+        the regression parameter used in Ridge and Lasso regression
+
+    regression : ["ridge", "lasso", "lstsq"], default "ridge"
+        the regression method if the number of points and terms don't match
+    """
+
+    def __init__(self, domain, basis_sets, alpha=1e-10, regression="ridge"):
+        super().__init__(domain)
+        self._basis_sets = None
+        self._alpha = None
+        self._regression = None
+        self._index_combinations = None
+        self._coefficients = None
+        self._fit_gradient_flag = False
+        self._integration_constant_flag = False
+
+        self._basis_sets = basis_sets
+        self.alpha = alpha
+        self.regression = regression
+
+    @property
+    def alpha(self):
+        """float: constant of the L2 term"""
+        return self._alpha
+
+    @alpha.setter
+    def alpha(self, value):
+        if value < 0:
+            raise ValueError("Alpha term must be positive")
+        if self.alpha != value:
+            self._alpha = value
+            self._valid_cache = False
+
+    @property
+    def regression(self):
+        """float: constant of the L2 and L1 term"""
+        return self._regression
+
+    @regression.setter
+    def regression(self, value):
+        regression = str(value).casefold()
+        if not regression in ["ridge", "lasso", "lstsq"]:
+            raise ValueError("Regression must be ridge or lasso")
+        if self.regression != regression:
+            self._regression = regression
+            self._valid_cache = False
+
+    @property
+    def basis_sets(self):
+        """list of BasisFunctionSet: the set of basis functions for the terms."""
+        return self._basis_sets
+
+    @property
+    def index_combinations(self):
```

Comment:
This doesn't seem like it will need to be in the public interface. I would remove the property and use an internal variable.

Reply:
Do you want `_create_terms` to be called in the constructor or in fit? Public or not, this is the most convenient for `_create_terms` to be called, and it definitely shouldn't be called more than once.

Reply:
It should be called during fitting. You can use the same cache pattern in the fit methods.

Reply:
okay